finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,690 @@
+use crate::{
+    new_network::{
+        connection_manager::{Authentication, Multiaddr, SessionHandler},
+        DataCommand, PeerId, Protocol,
+    },
+    NodeIndex, SessionId,
+};
+use codec::{Decode, Encode};
+use std::{
+    collections::{HashMap, HashSet},
+    time,
+};
+
+/// Messages used for discovery and authentication.
+#[derive(Clone, Debug, Encode, Decode)]
+pub enum DiscoveryMessage {
+    // Contains the id of the broadcast, to avoid spam.
+    AuthenticationBroadcast(Authentication),
+    // Requests always contain own authentication, to avoid asymetric trust.
+    Request(Vec<NodeIndex>, Authentication),
+    Authentications(Vec<Authentication>),
+}
+
+impl DiscoveryMessage {
+    pub fn session_id(&self) -> SessionId {
+        use DiscoveryMessage::*;
+        match self {
+            AuthenticationBroadcast((auth_data, _)) => auth_data.session(),
+            Request(_, (auth_data, _)) => auth_data.session(),
+            Authentications(auths) => match auths.get(0) {
```

Comment:
as far as I understand, we assume that all the authentications come from the same session; I would add either a check or at least a comment near the variant

Reply:
Added comment to the variant.